airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/java/io/airbyte/cdk/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -215,6 +217,45 @@ private String loadDataIntoBucket(final String objectPath, final SerializableBuf
     return newFilename;
   }
 
+  /**
+   * Users want deterministic file names (e.g. the first file part is really foo-0.csv Using UUIDs
+   * (previous approach) doesn't allow that. However, using pure integers could lead to a collision
+   * with an existing file. So, we'll count up the existing files in the directory and use that as a
+   * lazy-offset, assuming airbyte manages the dir and has similar naming conventions. `getPartId`
+   * will be 1-indexed.
+   */
+  @VisibleForTesting
+  String getPartId(String objectPath) {
+    if (partCounts.get(objectPath) == null) {
+      partCounts.put(objectPath, new AtomicInteger(0));
+    }
+
+    final AtomicInteger partCount = partCounts.get(objectPath);
+
+    if (partCount.get() == 0) {
```

Comment:
what's this check intended to handle? (plausibly I'm misunderstanding)

generally we recommend that users include a timestamp in their output path, so we shouldn't overwrite files from a previous sync. And presumably, if they _don't_ do that, then their intent is that we _do_ overwrite previous files (... and for some reason don't want to use the overwrite mode)

and if we do the partCount stuff correctly, then different threads should never collide on the same partId value

Reply:
I was looking at the old implementation before the UUIDs (https://github.com/airbytehq/airbyte/pull/29640/files), and the code went through this rigmarole which I *think* had the intent of continuing off where the previous part numbers stopped.  This makes sense to me in a multi-attempt sync:   
* Attempt 1: Files `bla-{1-3}.csv`
* Attempt 2: Files `bla-{4-6}.csv`
* etc.

Since one of the complaints against the UUID approach is that folks want deterministic file names (https://github.com/airbytehq/airbyte/pull/32228#issue-1980201458), I think bringing this back is *probably* a good idea?

Reply:
Whoa, I had no idea. I guess we should keep that then. Nice find!

one last question - do we have a list of object paths at the start of the sync? I.e. could we do this as a one-time setup, instead of listing objects every time we make a new file.

And if not - this logic definitely needs to be threadsafe, since it's doing IO (and therefore very likely to race). Something like this, I think?
```java
// new field for locking
private static final Object OBJECT_LISTING_LOCK = new Object(); // could be per-objectPath, I'm being lazy

// and then inside the method
if (partCount.get() == 0) {
  synchronized (OBJECT_LISTING_LOCK) {
    // Check if another thread got the lock first
    if (partCount.get() == 0) {
      final String bucket = s3Config.getBucketName();
      final ObjectListing objects = s3Client.listObjects(bucket, objectPath);
      // etc...

      // return from inside the synchronized block to avoid race condition
      // where another thread gets the lock+passes its if(partCount == 0) check
      // before this method increments the counter
      // This can happen inside the if-block, because if we skip the object listing,
      // then we're not mutating any state (so we don't need to hold the lock while we getAndIncrement)
      partCount.set(whatever + 1);
      return whatever;
    }
  }
}

return partCount.getAndIncrement();
```

also - how does this even work? I don't think we guarantee identical timestamps across attempts (in fact, I would have assumed we explicitly _don't_) - so different attempts shouldn't be writing to the same files to begin with. Maybe this is for users who use filepaths _without_ timestamps? ... but then we could _easily_ exceed 100 files in a single directory, and the next sync will run into the exact same problem (at which point, having a hardcoded `100` will still lead to data loss...)

AWS says they include a [pagination token](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#AmazonS3-ListObjectsV2-response-NextContinuationToken) in the response, can we use that to get the full listing?

Reply:
Yep, I guess users will need to change the default naming scheme to remove the epoch to resume partIDs, but it's possible!  Ideally that would be a sync-id one day anyway... 

I figure at this point we might as well make the whole method `synchonized` and then we can recuse though the S3 listing.  - done in 1549fe6e944f4bf30e5b421aef29c1dedf052d0f